File: autojs/src/main/java/com/stardust/autojs/rhino/continuation/Continuation.kt
package com.stardust.autojs.rhino.continuation

import com.stardust.autojs.core.looper.Timer
import com.stardust.autojs.core.looper.TimerThread
import com.stardust.autojs.rhino.AutoJsContext
import com.stardust.autojs.runtime.ScriptRuntime

import org.mozilla.javascript.Context
import org.mozilla.javascript.ContinuationPending
import org.mozilla.javascript.Scriptable

class Continuation(val context: AutoJsContext, val scope: Scriptable, private val mTimer: Timer) {
    var pending: ContinuationPending? = null
        private set
    private val mThread: Thread = Thread.currentThread()

    class Result(val result: Any?, val error: Any?) {
        companion object {

            fun success(result: Any?): Result {
                return Result(result, null)
            }

            fun failure(error: Any?): Result {
                return Result(null, error)
            }
        }
    }

    fun suspend() {
        if (pending != null) {
            throw IllegalStateException("call suspend twice!")
        }
        context.captureContinuation().let {
            pending = it
            throw it
        }
    }

    fun resumeWith(result: Result) {
        val continuation = pending?.continuation
                ?: throw IllegalStateException("call resume() without suspend()!")
        if (mThread == Thread.currentThread()) {
            context.resumeContinuation(continuation, scope, result)
        } else {
            mTimer.postDelayed({
                context.resumeContinuation(continuation, scope, result)
            }, 0)
        }
    }

    companion object {

        fun create(runtime: ScriptRuntime, scope: Scriptable): Continuation {
            val context = Context.getCurrentContext() as AutoJsContext
            return Continuation(context, scope, runtime.timers.timerForCurrentThread)
        }
    }
}

File: autojs/src/main/assets/modules/__continuation__.js
module.exports = function (runtime, global) {
    const Continuation = com.stardust.autojs.rhino.continuation.Continuation;

    function continuation() {

    }

    continuation.create = function (scope) {
        scope = scope || global;
        var cont = Object.create(runtime.createContinuation(scope));
        cont.await = function () {
            let result = cont.suspend();
            if (result.error != null) {
                throw result.error;
            }
            return result.result;
        };
        cont.resumeError = function (error) {
            if (error == null || error == undefined) {
                throw TypeError("error is null or undefined");
            }
            cont.resumeWith(Continuation.Result.Companion.failure(error));
        }
        cont.resume = function (result) {
            cont.resumeWith(Continuation.Result.Companion.success(result));
        }
        return cont;
    }

    function awaitPromise(scope, promise) {
        var cont = continuation.create(scope);
        promise.then(result => {
            cont.resume(result);
        }).catch(error => {
            cont.resumeError(error);
        });
        return cont.await();
    }

    continuation.await = function (any) {
        if (Object.getPrototypeOf(any).constructor === Promise) {
            return awaitPromise(global, any);
        }
        throw new TypeError('cannot await ' + any);
    }

    continuation.delay = function (millis) {
         var cont = continuation.create();
         setTimeout(()=>{
             cont.resume();
         }, millis);
         cont.await();
     }

    continuation.__defineGetter__('enabled', function () {
        return engines.myEngine().hasFeature("continuation");
    });

    global.Promise.prototype.await = function () {
        return continuation.await(this);
    }

    return continuation;
}